core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java

```diff
@@ -87,6 +108,34 @@ public static BaseFileSystem create(FileSystemContext context) {
    */
   protected BaseFileSystem(FileSystemContext fsContext) {
     mFsContext = fsContext;
+    mBlockStore = AlluxioBlockStore.create(fsContext);
+  }
+
+  /**
+   * Shuts down the FileSystem. Closes all thread pools and resources used to perform operations. If
+   * any operations are called after closing the context the behavior is undefined.
+   *
+   * @throws IOException
+   */
+  @Override
+  public void close() throws IOException {
+    if (!mClosed) {
+      synchronized (mCloseLock) {
+        // TODO(zac) Determine the behavior when closing the context during operations.
+        if (!mClosed) {
+          mClosed = true;
+          Factory.FILESYSTEM_CACHE.remove(new FileSystemKey(mFsContext.getClientContext()));
```

Comment:
Could this cause problems if we create two filesystems for the same scheme/authority, one with custom configuration, then close the filesystem with custom configuration? The key only uses scheme/authority, so closing the non-cached filesystem will remove the cached filesystem.

Reply:
if you use custom configuration you forgo caching, so that wouldn't be an issue.

Reply:
example test
```java
  @Test
  public void dontRemoveUncachedFilesystems() throws Exception {
    FileSystem fs1 = FileSystem.Factory.get();
    InstancedConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
    conf.set(PropertyKey.USER_WORKER_LIST_REFRESH_INTERVAL, "1sec");
    FileSystem fs2 = FileSystem.Factory.create(conf);
    fs2.close();
    FileSystem fs3 = Factory.get();
    assertSame(fs1, fs3);
  }
```